servant-server/src/Servant/Server/Internal.hs

```diff
@@ -852,5 +856,53 @@ type HasServerArrowTypeError a b =
     ':$$: 'ShowType b
 #endif
 
+-- | If you use @'Fragment' Text@ in one of the endpoints for your API,
+-- this automatically requires your server-side handler to be a function
+-- that takes an argument of type @'Maybe' 'Text'@.
+--
+-- This lets servant worry about looking it up in the URI fragment
+-- and turning it into a value of the type you specify, enclosed
+-- in 'Maybe', because it may not be there and servant would then
+-- hand you 'Nothing'.
+--
+-- You can control how it'll be converted from 'Text' to your type
+-- by simply providing an instance of 'FromHttpApiData' for your type.
+--
+-- Example:
+--
+-- > type MyApi = "books" :> Fragment Text :> Get '[JSON] [Book]
+-- >
+-- > server :: Server MyApi
+-- > server = getBooksBy
+-- >   where getBooksBy :: Maybe Text -> Handler [Book]
+-- >         getBooksBy Nothing       = ...return all books...
+-- >         getBooksBy (Just author) = ...return books by the given author...
+#ifdef HAS_TYPE_ERROR
+instance (OnlyOneFragment api, HasServer api context, FromHttpApiData a1 )
+#else
+instance (HasServer api context, FromHttpApiData a1 )
+#endif
+    => HasServer (Fragment a1 :> api) context where
+  type ServerT (Fragment a1 :> api) m = Maybe a1 -> ServerT api m
+
+  route Proxy context subserver =
+    let parseFragment = join . fmap (either (const Nothing) Just)
+          . join . fmap (either (const Nothing) Just)
+          . fmap (fmap parseQueryParam . T.decodeUtf8')
+          . identifyFragment (not . BC8.null) (crop . rawPathInfo) (crop . rawQueryString)
```

Comment:
I think instead of `(not . BC8.null)` we should check the presence of `#`. The difference is that `/my/endpoint#` will have an empty fragment while `/my/endpoint` will have no fragment.

Reply:
Fixed.